### ISSUE TITLE
ignoring corrupt image/label: Label class 9 exceeds dataset class count 2. Possible class labels are 0-1

### ISSUE THREAD

**Issue by ddelikonstantis** [ultralytics/ultralytics#23687]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello everyone. 

I have a custom dataset with a data.yaml that has 10 classes. It looks like this.

names:
  0: class_1
  1: class_2
  2: class_3
  3: class_4
  4: class_5
  5: class_6
  6: class_7
  7: class_8
  8: class_9
  9: class_10
path: /var/folders/ln/
train: images/train/
val: images/val/

I want to be able to train selectively on subsets of the above classes.

For example, I want to train on:

3: class_4
7: class_8

Furthermore, I want to fine-tuning on my custom dataset. For example,
1st training: Train for 50 epochs on 
  7: class_8
  8: class_9
2nd training: Train for 50 epochs on 
  5: class_6
  6: class_7

and so on.

How can I do that?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ddelikonstantis, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

From your error message (`Label class 15 exceeds dataset class count 2...`) and your goal of training on class subsets (and doing staged fine-tuning), it would really help to see exactly how the dataset YAML + labels + training command are being set up for each stage 🙂.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.  
-Please include: your exact training command(s), the `data.yaml` you used for that run, and 1–2 example label `.txt` files that trigger the warning (plus the corresponding image filenames if possible).

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response 🤖✅, and an Ultralytics engineer will also assist soon.

**Comment by onuralpszr** [comment]
That error is a classic dataloader mismatch. When you manually edit the names list in your data.yaml to include only a subset of classes (like 2 classes), the dataloader resets the maximum expected class index to 1. However, your yaml label files still contain the original absolute indices (like 3, 7, 8, or 9). When the YOLO dataset parser reads a label index that falls outside the 0 to len(names)-1 range, it flags the label as corrupt and skips the image entirely.

The cleanest way to handle this without rewriting your .yml files or creating multiple fragmented data.yaml files is to use the native classes argument during training.

Here is how you can set it up.

1. Keep Your Original data.yaml
Leave your data.yaml exactly as it is, with all 10 classes defined.

```yaml
names:
  0: class_1
  1: class_2
  2: class_3
  3: class_4
  4: class_5
  5: class_6
  6: class_7
  7: class_8
  8: class_9
  9: class_10
path: /var/folders/ln/
train: images/train/
val: images/val/
```


2. Use the classes Argument for Selective Training

When you initialize the training run, simply pass a list of the class indices you want to filter for. The dataloader will automatically ignore the bounding boxes for unselected classes and map your selected classes correctly in the background.

```python
from ultralytics import YOLO

# Initialize a base model
model = YOLO('yolo26n.pt') 

# Train ONLY on class_4 (index 3) and class_8 (index 7)
model.train(data='data.yaml', classes=[3, 7], epochs=50)
```

**Comment by ddelikonstantis** [comment]
> That error is a classic dataloader mismatch. When you manually edit the names list in your data.yaml to include only a subset of classes (like 2 classes), the dataloader resets the maximum expected class index to 1. However, your yaml label files still contain the original absolute indices (like 3, 7, 8, or 9). When the YOLO dataset parser reads a label index that falls outside the 0 to len(names)-1 range, it flags the label as corrupt and skips the image entirely.
> 
> The cleanest way to handle this without rewriting your .yml files or creating multiple fragmented data.yaml files is to use the native classes argument during training.
> 
> Here is how you can set it up.
> 
>     1. Keep Your Original data.yaml
>        Leave your data.yaml exactly as it is, with all 10 classes defined.
> 
> 
> names:
>   0: class_1
>   1: class_2
>   2: class_3
>   3: class_4
>   4: class_5
>   5: class_6
>   6: class_7
>   7: class_8
>   8: class_9
>   9: class_10
> path: /var/folders/ln/
> train: images/train/
> val: images/val/
> 
>     2. Use the classes Argument for Selective Training
> 
> 
> When you initialize the training run, simply pass a list of the class indices you want to filter for. The dataloader will automatically ignore the bounding boxes for unselected classes and map your selected classes correctly in the background.
> 
> from ultralytics import YOLO
> 
> # Initialize a base model
> model = YOLO('yolo26n.pt') 
> 
> # Train ONLY on class_4 (index 3) and class_8 (index 7)
> model.train(data='data.yaml', classes=[3, 7], epochs=50)

Thank you for your answer. I have a few more questions.

1. If I would like to do fine-tuning on a dataset where the first training has classes [0, 7, 15] and then do fine-tuning on a dataset with classes [1, 2, 3], then I guess it is important to have the original data.yaml defined with all classes as you suggested, correct?
2. Is there a way to keep the original data.yaml and treat all the classes as a single class?

**Comment by Y-T-G** [comment]
1. Yes, the original YAML needs to have all the classes defined. However, fine-tuning like this most of the time doesn't work. The model simply will get worse on the other classes. But if you mean these 3 are the same classes, but with different indices, then there isn't a way to fine-tune like that. Because each class index is treated as a different class. So [0,1,2] would be different class from [4,5,6] even if you fine-tune separately. The model will output 6 classes, not 3. You would have to make the class indices the same in the labels if you want to treat them as same class.
2. You can use `single_cls=True` to train all the classes as the same class.
